### 1. Đọc bộ dữ liệu, cho biết số dòng, số cột và kiểu dữ liệu của các thuộc tính

In [1]:
import pandas as pd
import numpy as np
df = pd.read_csv('OnlineRetail.csv', encoding = 'ISO-8859-1')
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


+ Dữ liệu có 541909 dòng, 8 cột
+ InvoiceNo, StockCode, Description, InvoiceDate, CustomerID, Country: thang đo định danh
+ Quantity, Unit Price: thang đo tỉ lệ

### 2. Xây dựng bảng Pivot Table, với mỗi Số hoá đơn tính trung bình cộng số lượng các mặt hàng theo từng Quốc gia

In [3]:
# đổi tên cột sang tiếng việt
df.rename(columns = {'InvoiceNo':'Sohoadon','StockCode':'Macophieu','Description':'Mota','InvoiceDate':'Ngay','CustomerID':'Makhachhang','Country':'Nuoc','Quantity':'Soluong','UnitPrice':'Gia'},inplace = True)
df.head()

Sohoadon Macophieu                                 Mota  Soluong  \
0   536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER        6   
1   536365     71053                  WHITE METAL LANTERN        6   
2   536365    84406B       CREAM CUPID HEARTS COAT HANGER        8   
3   536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE        6   
4   536365    84029E       RED WOOLLY HOTTIE WHITE HEART.        6   

             Ngay   Gia  Makhachhang            Nuoc  
0  12/1/2010 8:26  2.55      17850.0  United Kingdom  
1  12/1/2010 8:26  3.39      17850.0  United Kingdom  
2  12/1/2010 8:26  2.75      17850.0  United Kingdom  
3  12/1/2010 8:26  3.39      17850.0  United Kingdom  
4  12/1/2010 8:26  3.39      17850.0  United Kingdom

In [4]:
df.pivot_table(values = 'Soluong', index = 'Sohoadon', columns = 'Nuoc', aggfunc = 'mean')

Nuoc      Australia  Austria  Bahrain  Belgium  Brazil  Canada  \
Sohoadon                                                         
536365          NaN      NaN      NaN      NaN     NaN     NaN   
536366          NaN      NaN      NaN      NaN     NaN     NaN   
536367          NaN      NaN      NaN      NaN     NaN     NaN   
536368          NaN      NaN      NaN      NaN     NaN     NaN   
536369          NaN      NaN      NaN      NaN     NaN     NaN   
...             ...      ...      ...      ...     ...     ...   
C581484         NaN      NaN      NaN      NaN     NaN     NaN   
C581490         NaN      NaN      NaN      NaN     NaN     NaN   
C581499         NaN      NaN      NaN      NaN     NaN     NaN   
C581568         NaN      NaN      NaN      NaN     NaN     NaN   
C581569         NaN      NaN      NaN      NaN     NaN     NaN   

Nuoc      Channel Islands  Cyprus  Czech Republic  Denmark  ...  RSA  \
Sohoadon                                                    ...        
536365                NaN     NaN             NaN      NaN  ...  NaN   
536366                NaN     NaN             NaN      NaN  ...  NaN   
536367                NaN     NaN             NaN      NaN  ...  NaN   
536368                NaN     NaN             NaN      NaN  ...  NaN   
536369                NaN     NaN             NaN      NaN  ...  NaN   
...                   ...     ...             ...      ...  ...  ...   
C581484               NaN     NaN             NaN      NaN  ...  NaN   
C581490               NaN     NaN             NaN      NaN  ...  NaN   
C581499               NaN     NaN             NaN      NaN  ...  NaN   
C581568               NaN     NaN             NaN      NaN  ...  NaN   
C581569               NaN     NaN             NaN      NaN  ...  NaN   

Nuoc      Saudi Arabia  Singapore  Spain  Sweden  Switzerland  USA  \
Sohoadon                                                             
536365             NaN        NaN    NaN     NaN          NaN  NaN   
536366             NaN        NaN    NaN     NaN          NaN  NaN   
536367             NaN        NaN    NaN     NaN          NaN  NaN   
536368             NaN        NaN    NaN     NaN          NaN  NaN   
536369             NaN        NaN    NaN     NaN          NaN  NaN   
...                ...        ...    ...     ...          ...  ...   
C581484            NaN        NaN    NaN     NaN          NaN  NaN   
C581490            NaN        NaN    NaN     NaN          NaN  NaN   
C581499            NaN        NaN    NaN     NaN          NaN  NaN   
C581568            NaN        NaN    NaN     NaN          NaN  NaN   
C581569            NaN        NaN    NaN     NaN          NaN  NaN   

Nuoc      United Arab Emirates  United Kingdom  Unspecified  
Sohoadon                                                     
536365                     NaN        5.714286          NaN  
536366                     NaN        6.000000          NaN  
536367                     NaN        6.916667          NaN  
536368                     NaN        3.750000          NaN  
536369                     NaN        3.000000          NaN  
...                        ...             ...          ...  
C581484                    NaN   -80995.000000          NaN  
C581490                    NaN      -11.500000          NaN  
C581499                    NaN       -1.000000          NaN  
C581568                    NaN       -5.000000          NaN  
C581569                    NaN       -3.000000          NaN  

[25900 rows x 38 columns]

### 3. Xây dựng bảng Pivot Table, với mỗi khách hàng cho biết số lượng mua hàng lớn nhất và nhỏ nhất theo Kho

In [5]:
df.pivot_table(values = 'Soluong', index = 'Macophieu', columns = 'Makhachhang', aggfunc = {min,max})

max                                                          \
Makhachhang 12346.0 12347.0 12348.0 12349.0 12350.0 12352.0 12353.0 12354.0   
Macophieu                                                                     
10002           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
10080           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
10120           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
10123C          NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
10124A          NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
...             ...     ...     ...     ...     ...     ...     ...     ...   
D               NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
DOT             NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
M               NaN     NaN     NaN     NaN     NaN     1.0     NaN     NaN   
PADS            NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
POST            NaN     NaN     6.0     1.0     1.0     2.0     NaN     NaN   

                             ...     min                                  \
Makhachhang 12355.0 12356.0  ... 18273.0 18274.0 18276.0 18277.0 18278.0   
Macophieu                    ...                                           
10002           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
10080           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
10120           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
10123C          NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
10124A          NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
...             ...     ...  ...     ...     ...     ...     ...     ...   
D               NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
DOT             NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
M               NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
PADS            NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
POST            NaN    18.0  ...     NaN     NaN     NaN     NaN     NaN   

                                                     
Makhachhang 18280.0 18281.0 18282.0 18283.0 18287.0  
Macophieu                                            
10002           NaN     NaN     NaN     NaN     NaN  
10080           NaN     NaN     NaN     NaN     NaN  
10120           NaN     NaN     NaN     NaN     NaN  
10123C          NaN     NaN     NaN     NaN     NaN  
10124A          NaN     NaN     NaN     NaN     NaN  
...             ...     ...     ...     ...     ...  
D               NaN     NaN     NaN     NaN     NaN  
DOT             NaN     NaN     NaN     NaN     NaN  
M               NaN     NaN     NaN     1.0     NaN  
PADS            NaN     NaN     NaN     NaN     NaN  
POST            NaN     NaN     NaN     NaN     NaN  

[3684 rows x 8744 columns]

### 4. Xây dựng bảng Pivot table, với mỗi Mã kho tính tổng số lượng các mặt hàng và trung bình cộng giá

In [6]:
df.pivot_table(values = ['Soluong','Gia'], index = 'Macophieu', aggfunc = {'Soluong':np.sum,'Gia':np.mean})

Gia  Soluong
Macophieu                       
10002          1.056849     1037
10080          0.376667      495
10120          0.210000      193
10123C         0.487500      -13
10123G         0.000000      -38
...                 ...      ...
gift_0001_20  15.038000       20
gift_0001_30  21.941250       37
gift_0001_40  33.566667        3
gift_0001_50  41.890000        4
m              2.550000        1

[4070 rows x 2 columns]

### 5. Xây dựng bảng Pivot table cho biết tổng số lượng hàng bán được của mỗi ngày

In [21]:
# lọc ngày
a = [x[:9] for x in df['Ngay']]
df['Ngay']= a
df.head()

Sohoadon Macophieu                                 Mota  Soluong       Ngay  \
0   536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER        6  12/1/2010   
1   536365     71053                  WHITE METAL LANTERN        6  12/1/2010   
2   536365    84406B       CREAM CUPID HEARTS COAT HANGER        8  12/1/2010   
3   536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE        6  12/1/2010   
4   536365    84029E       RED WOOLLY HOTTIE WHITE HEART.        6  12/1/2010   

    Gia  Makhachhang            Nuoc  
0  2.55      17850.0  United Kingdom  
1  3.39      17850.0  United Kingdom  
2  2.75      17850.0  United Kingdom  
3  3.39      17850.0  United Kingdom  
4  3.39      17850.0  United Kingdom

In [22]:
df.pivot_table(values = 'Soluong', index = 'Ngay', aggfunc = 'sum')

Soluong
Ngay              
1/10/2011     9014
1/11/2011    29093
1/12/2011    10133
1/13/2011    10114
1/14/2011    23091
...            ...
9/5/2011     23484
9/6/2011     14959
9/7/2011     16211
9/8/2011     16944
9/9/2011     16100

[305 rows x 1 columns]